### PR TITLE
refactor(connectors:sqla): break sqla query generator

### PR DESCRIPTION
**Pull request by mayurnewase**
### SUMMARY

This is the first part of many upcoming pr to clean and break sql query generator.
The original [PR](https://github.com/apache/superset/pull/12482), needed to be broken down due to many conflicts,so this pr separates first few functions.

### TEST PLAN

1.Run all existing and newly added tests.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by mayurnewase** [comment]
@villebro

**Comment by mayurnewase** [comment]
@villebro

**Comment by ktmud** [comment]
Hi, @mayurnewase , thank you for iterating on this. Could you complete the PR template and make sure to add a link to your [previous PR](https://github.com/apache/superset/pull/12482) for context?

I know it feels laborious and maybe meaningless, but following the PR template makes sure each PR is easily reviewable. You can just fill in `N/A` if some fields do not apply.

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -900,20 +924,40 @@ def get_sqla_query(  # pylint: disable=too-many-arguments,too-many-locals,too-ma
...
-            "row_offset": row_offset,
-            "to_dttm": to_dttm.isoformat() if to_dttm else None,
-            "filter": filter,
-            "columns": [col.column_name for col in self.columns],
-        }
-        template_kwargs.update(self.template_params_dict)
+
         extra_cache_keys: List[Any] = []
-        template_kwargs["extra_cache_keys"] = extra_cache_keys
+        is_sip_38 = is_feature_enabled("SIP_38_VIZ_REARCHITECTURE")
```
Comment (villebro) [inline superset/connectors/sqla/models.py:929]: This feature flag has been removed a while ago, so you probably need to double check that all changes here are still relevant.

File: superset/connectors/sqla/models.py [hunk 2]
```diff
@@ -877,6 +878,29 @@ def _get_sqla_row_level_filters(
...
+            "groupby": groupby,
+            "metrics": metrics,
+            "row_limit": row_limit,
+            "row_offset": row_offset,
+            "to_dttm": to_dttm.isoformat() if to_dttm else None,
+            "filter": filter_,
+            "columns": [col.column_name for col in self.columns],
+            "extra_cache_keys": extra_cache_keys,
+        }
+
```
Comment (villebro) [inline superset/connectors/sqla/models.py:903]: Would it make sense to combine `_get_template_kwargs` and `update_template_kwargs` into one function in `sql_query_utils.py`?

File: superset/connectors/sqla/sql_query_utils.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import Any, Dict
+
+
+def update_template_kwargs(
```
Comment (willbarrett) [inline superset/connectors/sqla/sql_query_utils.py:20]: Is there value in pulling out this function if it's a 1-liner that runs `update` on the first dict with the second?

**Comment by junlincc** [comment]
@mayurnewase Thank you Mayur for contribution and collaboration!🙏 please do add more description for project tracking and release purposes.

**Review by ktmud — APPROVED** [review]

**Review by ktmud — COMMENTED** [review]
Sorry, hit Approve by mistake.

I think we need to think deeper on how to structure the SQL query generator in an extensible and maintainable way. There's nothing wrong with splitting things into various smaller functions, but considering how many variables we need to manipulate and manage, a `QueryBuilder` class might make more sense. There's definitely room for util functions, but it is better to make sure they only accept a couple of arguments (<= 4), so we don't have to pass a lot of parameters around.

**Comment by mayurnewase** [comment]
> Sorry, hit Approve by mistake.
> 
> I think we need to think deeper on how to structure the SQL query generator in an extensible and maintainable way. There's nothing wrong with splitting things into various smaller functions, but considering how many variables we need to manipulate and manage, a `QueryBuilder` class might make more sense. There's definitely room for util functions, but it is better to make sure they only accept a couple of arguments (<= 4), so we don't have to pass a lot of parameters around.

Yeah,I went through how sqlalchemy uses `Query` class to generate queries which makes it easier.
I will try to do in that fashion in future prs if required.

**Pull request closed without merging**